src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,732 @@
+    if array.Length = 0 then None
+    else Some array.[0]
+
+let tail (array : 'T[]) =
+    if array.Length = 0 then invalidArg "array" "Not enough elements"
+    slice array 1
+
+let item index (array: _[]) =
+    array.[index]
+
+let tryItem index (array:'T[]) =
+    if index < 0 || index >= array.Length then None
+    else Some array.[index]
+
+let toList array = 
+    List.ofArray array
+
+
+// TODO: Consider optimizing functions below via manual inlining
+// -or- re-implementing via native js arrays functions (i.e. reduce(...))
+
+let foldBackIndexed<'T,'State> folder (array: 'T[]) (state:'State) =
+   let mutable acc = state
+   let size = length array
+   for i = 1 to size do
+      acc <- folder (i-1) array.[size - i] acc
+   acc
+
+let foldBack<'T,'State> folder (array: 'T[]) (state:'State) =
+   foldBackIndexed (fun _ x acc -> folder x acc) array state
+
+let foldIndexed2 folder state array1 array2 =
+   let mutable acc = state
+   if length array1 <> length array2 then failwith "Arrays have different lengths"
+   for i = 0 to length array1 - 1 do
+      acc <- folder i acc array1.[i] array2.[i]
+   acc
+
+let fold2<'T1, 'T2, 'State> folder (state: 'State) (array1: 'T1[]) (array2: 'T2[]) =
+   foldIndexed2 (fun _ acc x y -> folder acc x y) state array1 array2
+
+let foldBackIndexed2<'T1, 'T2, 'State> folder (array1: 'T1[]) (array2: 'T2[]) (state:'State) =
+   let mutable acc = state
+   if length array1 <> length array2 then failwith "Arrays had different lengths"
+   let size = length array1
+   for i = 1 to size do
+      acc <- folder (i-1) array1.[size - i] array2.[size - i] acc
+   acc
+
+let foldBack2<'T1, 'T2, 'State> f (array1: 'T1[]) (array2: 'T2[]) (state: 'State) = 
+   foldBackIndexed2 (fun _ x y acc -> f x y acc) array1 array2 state 
+
+let reduce reduction array =
+   if length array = 0 then invalidOp "Array was empty"
+   else foldIndexed (fun i acc x -> if i = 0 then x else reduction acc x) Unchecked.defaultof<_> array
+
+let reduceBack reduction array =
+   if length array = 0 then invalidOp "Array was empty"
+   else foldBackIndexed (fun i x acc -> if i = 0 then x else reduction acc x) array Unchecked.defaultof<_>
+
+let forAll2 predicate array1 array2 =
+   fold2 (fun acc x y -> acc && predicate x y) true array1 array2
+
+let rec existsOffset predicate array index =
+   if index = length array then false
+   else predicate array.[index] || existsOffset predicate array (index+1)
+
+let exists predicate array = 
+   existsOffset predicate array 0
+
+let rec existsOffset2 predicate array1 (array2:_ []) index =
+   if index = length array1 then false
+   else predicate array1.[index] array2.[index] || existsOffset2 predicate array1 array2 (index+1)
+
+let rec exists2 predicate array1 array2 =
+   if length array1 <> length array2 then failwith "Arrays had different lengths"
+   existsOffset2 predicate array1 array2 0
+
+let inline sum (array: 'T []) : 'T =
+   fold (+) LanguagePrimitives.GenericZero array
```

Comment:
Unfortunately we cannot use `inline` in public methods here because Fable would erase them. Currently fable-core is distributed in JS form and Fable won't have access to the F# source to read the inlined expression.

Reply:
Oh, indeed. Got it!